lib/active_merchant/billing/gateways/cecabank/cecabank_json.rb

```diff
@@ -0,0 +1,231 @@
+      end
+
+      def void(identification, options = {})
+        authorization, operation_number, money, _network_transaction_id = identification.split('#')
+        options[:operation_number] = operation_number
+        build_cancellation(:void, money.to_i, authorization, options)
+      end
+
+      def refund(money, identification, options = {})
+        authorization, operation_number, _money, _network_transaction_id = identification.split('#')
+        options[:operation_number] = operation_number
+        build_cancellation(:refund, money, authorization, options)
+      end
+
+      private
+
+      def build_purchase(action, money, creditcard, options)
+        post = { parametros: { accion: CECA_ACTIONS_DICTIONARY[action] } }
+
+        add_invoice(post, money, options)
+        add_creditcard(post, creditcard)
+        add_stored_credentials(post, creditcard, options)
+        add_three_d_secure(post, options)
+
+        commit('compra', post)
+      end
+
+      def build_cancellation(action, money, authorization, options = {})
+        post = {}
+        add_auth_invoice_data(action, post, money, authorization, options)
+
+        commit('anulacion', post)
+      end
+
+      def add_auth_invoice_data(action, post, money, authorization, options)
+        params = post[:parametros] ||= {}
+        params[:accion] = CECA_ACTIONS_DICTIONARY[action]
+        params[:referencia] = authorization
+
+        add_invoice(post, money, options)
+      end
+
+      def add_encryption(post)
+        post[:cifrado] = CECA_ENCRIPTION
+      end
+
+      def add_signature(post, params_encoded, options)
+        post[:firma] = Digest::SHA2.hexdigest(@options[:signature_key].to_s + params_encoded)
+      end
+
+      def add_merchant_data(post)
+        params = post[:parametros] ||= {}
+
+        params[:merchantID] = @options[:merchant_id]
+        params[:acquirerBIN] = @options[:acquirer_bin]
+        params[:terminalID] = @options[:terminal_id]
+      end
+
+      def add_invoice(post, money, options)
+        post[:parametros][:numOperacion] = options[:operation_number] || options[:order_id]
+        post[:parametros][:importe] = amount(money)
+        post[:parametros][:tipoMoneda] = CECA_CURRENCIES_DICTIONARY[options[:currency] || currency(money)].to_s
+        post[:parametros][:exponente] = 2.to_s
+      end
+
+      def add_creditcard(post, creditcard)
+        params = post[:parametros] ||= {}
+
+        params[:pan] = creditcard.number
+        params[:caducidad] = strftime_yyyymm(creditcard)
+        params[:cvv2] = creditcard.verification_value
+        params[:csc] = creditcard.verification_value if CreditCard.brand?(creditcard.number) == 'american_express'
+      end
+
+      def add_stored_credentials(post, creditcard, options)
+        return unless stored_credential = options[:stored_credential]
+
+        return if options[:exemption_sca] == 'NONE' || options[:exemption_sca].blank?
```

Comment:
❓ Question:

If we already have info from the stored_credentials or from the [parameters already](https://docs.spreedly.com/guides/spreedly-3dsecure2/#requesting-exemptions) in place for exemptions maybe we don't need a GSF here ?

Reply:
We will need this `exemption_sca` field because it is a different kind of exemption that describe the guide, also `exemption_sca` related the Cecabank guide includes another king of values (not only to trigger MIT transactions and with another additional purpose of marking the transactions as low_value_exemption for 3DS).